### ISSUE TITLE
SCLP HWL / Store Data compatibility with zLinux

### ISSUE THREAD

**Issue by bluecmd** [SDL-Hercules-390/hyperion#185]
Hi,

First: I'm very much a newbie on the S390/Z architecture so forgive me if I do something stupid.

Now, something that caught my attention is the loader facility that Hercules calls HWL but IBM appears to be calling Store Data (as evident from [this commit](https://github.com/torvalds/linux/commit/66aec647216f129b8560dba738303a8486481c53#diff-9eeeb1dc96af1958582357af6be8eae7)).

The interface seems to provide blob access to what is referred to as "files". The ones I know of are:

- File 1: Coupling Facility Control Code (CFCC)
- File 2: SCSI Boot Loader
- File 3: Configuration of some sort (that's what the kernel commit above adds)
- File 7: Netboot (Unconfirmed, the only reference to this I find is [in Hercules](https://github.com/SDL-Hercules-390/hyperion/blob/54070a2f353313342d973d12c36425798d4e7dee/scescsi.c#L113))

As a side note I have reached out to the author of that patch set with the suggestion to add the other file types as well as it doesn't make much sense not to expose them I feel.

As I'm a curious person I wanted to see if I could use the SCLP SD interface contributed above to read an arbitrary file to an Ubuntu S390X VM running under Hercules. Turns out, I cannot. Which probably means that Hercules is not emulating something perfectly, or the Linux implementation is flawed. Whichever way, it's a nice win to figure it out.

# What's the environment?

- I run Ubuntu 18.04 Bionic with the latest 4.18 kernel. After installation it comes with 4.15 so it needs to be updated (`sudo apt-get install linux-image-4.18.0-15-generic`) and re-IPLd.
- I'm running a couple of days old Hyperion (7703dda3) with some custom tracing printouts added.

I have applied this patch to `scescsi.c` (manually cleaned up so it will not apply using `patch` but you get the idea):

```diff
diff --git a/scescsi.c b/scescsi.c
index 1754889b..29d1e423 100644
--- a/scescsi.c
+++ b/scescsi.c
@@ -89,7 +89,9 @@ typedef struct _SCCB_HWL_BK {
 #define SCCB_HWL_TYPE_RESET     0x01    /* Reset request             */
 #define SCCB_HWL_TYPE_INFO      0x02    /* Load info request         */
 /*007*/ BYTE    file;
 #define SCCB_HWL_FILE_SCSIBOOT  0x02    /* SCSI Boot Loader          */
+#define SCCB_HWL_FILE_CONFIG    0x03    /* Config Loader             */
 /*008*/ FWORD   resv1[2];
 /*010*/ FWORD   hwl;                    /* Pointer to HWL BK         */
 /*014*/ HWORD   resv2;
@@ -108,7 +110,9 @@ struct name2file {
 
 
 struct name2file n2flist[] = {
     { "scsiboot", SCCB_HWL_FILE_SCSIBOOT },
+    { "config",   SCCB_HWL_FILE_CONFIG },
 #if 0
     { "netboot",  7 }, // Hercules implementation of PXE style netboot for OSA
 #endif
```

[This ](https://gist.github.com/bluecmd/9490baa2a1b3c2a896e05083cdd45446) is my configuration for Hyperion.

`/usr/local/share/hercules/scsiboot` is the same file as `util/scsiboot` in this repository (hash checksum verified). The size is 4096 bytes.

# What's the issue 

This is how I trigger the issue from inside the Ubuntu VM:

```
root@ubuntu:/sys/firmware/sclp_sd/config# echo 1 > reload
root@ubuntu:/sys/firmware/sclp_sd/config# stat data
  File: data
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: 14h/20d	Inode: 7271        Links: 1
Access: (0444/-r--r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2019-02-11 22:44:41.444374503 +0100
Modify: 2019-02-11 22:44:41.444374503 +0100
Change: 2019-02-11 22:44:41.444374503 +0100
 Birth: -
root@ubuntu:/sys/firmware/sclp_sd/config# dmesg | tail -n 1
[ 2678.416112] sclp_sd.aa051d: A 4096-byte config data entity was retrieved
root@ubuntu:/sys/firmware/sclp_sd/config# hexdump -C data
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00001000
root@ubuntu:/sys/firmware/sclp_sd/config# 
```

As you can see the size is transmitted successfully, However the file is only zeros inside the VM so no data was copied..

From my [trace output](https://gist.github.com/bluecmd/2762655d3486b8160fe7210bf063466b) it can be seen that Hercules tried to honor the copy but failed. My modifications to the `hwl_loadfile` function is included in that link as well.

Note this part:

```
TRACE pto: 24     
```
(That's 0x24, not decimal)

This means that the check for PAGETAB_INVALID (0x20) fails. I'm not sure why this page table would be marked as INVALID or what this means here.

# Debugging

PTO is constructed from what I can deduce is dereferencing STO, which is called SAT in the Linux code. Here is my very professional translation table to map what is what compared to Hercules' `_SCCB_HWL_BK`:

```c
struct sclp_sd_evbuf {                 
  ·struct evbuf_header hdr;             
  u8 eq; /* type */                    
  u8 di; /*file */                     
  u8 rflags; /* resv1[0] */               
  u64 :56; /* resv1[0] also */        
  u32 id; /* resv1[1] */               
  u16 :16; /* resv2 */                 
  u8 fmt; /* asa */                    
  u8 status; /* resv3 */               
  u64 sat; /* sto */                   
  u64 sa;  /* resv4[0] | resv4[1] */   
  u32 esize; /* resv4[2] */            
  u32 dsize; /* dsize */               
} __packed;    
```

In the relevant context SAT is called `asce` and is used as follows:

https://github.com/torvalds/linux/blob/66aec647216f129b8560dba738303a8486481c53/drivers/s390/char/sclp_sd.c#L309-L320

Following this deeper, the following function `base_asce_alloc` is called. Interestingly it appears to have been committed at the same time (Feb 22, 2018).

https://github.com/torvalds/linux/blob/aa0c38cf39de73bf7360a3da8f1707601261e518/arch/s390/mm/pgalloc.c#L630-L668

This function initializes the table with e.g. `_REGION3_ENTRY_EMPTY` for region 3 which just happens to evaluate to 0x24. So that's my current thinking is where the PTO value comes from.

So given the recent nature and the somewhat arcane uses of this config file, maybe this is a kernel bug?

I have it on good authority that IPLing SCSIBOOT from z/VM works when running under Hyperion, so this code is not totally broken - just this particular interaction seems to be a bit wonky.

A final note is that 4.18 from Ubuntu is of course not the same as 4.20 mainline, but compiling the Linux kernel in a VM running on Hercules is not exactly fast so I'm trying to figure out as much as I can without resorting to that.

**Comment by mfsysprog** [comment]
Not sure if it is any use in your case, but I needed this 'quick and dirty' patch to be able to succesfully load a file using the hardware loader:
diff --git a/scescsi.c b/scescsi.c
index 6ede9c40..eddbe478 100644
--- a/scescsi.c
+++ b/scescsi.c
@@ -776,6 +776,8 @@ int n;
                 if(hwl_fn[file])
                     free(hwl_fn[file]);
 
+               file = 1;
+
                 if(strcasecmp(argv[n],"none") && strlen(argv[n]))
                     hwl_fn[file] = strdup(argv[n]);
                 else

I can't remember or understand anymore why it made a difference, just know in my case it seems to work.

**Comment by ivan-w** [comment]
Hmmm...  Do you mean you are doing a SCSI boot over FCP (and btw, setting file to 1 sets the dump device filename) ?

Asking because the only path into hwldr_cmd() is via the hcpldr command.

and also... the load file filetype is filetype 0 !

--Ivan

**Comment by mfsysprog** [comment]
I don't recall anymore what I was trying to do. I'm starting to doubt myself if it actually did something useful, or that it was just trying out something and left it at some point. Sorry for the confusion.

**Comment by mfsysprog** [comment]
After my erronous comment I decided to try and figure out what is going on here. I've managed to recreate the situation following the OP's comments. What I found so far is that when you look at the sto value before the ```sto &= ASCE_TO;``` statement it has the value 7 at the end. This indicates that the linux kernel is passing a region 3 entry as set in the ```asce = table | _ASCE_TYPE_REGION3 | _ASCE_TABLE_LENGTH;``` statement from the base_asce_alloc function. If the logic in base_asce_alloc is correct, it should fill the region3 entry with a pointer to the segment table, which should point to the page table, which should point to the data page, but it seems that there is some logic error there so the entry remains at the 0x24 value. So problem one indeed seems to me to be a kernel bug.
I've also noticed using the R instruction to look at the memory that the 0x24 value sometimes changes after some time, but that could be because the memory is reallocated? It doesn't seem to be a pointer to a segment table anyways.
However, the hardware loader code also is not prepared to receive a region table, it assumes a segment table (for a 31 bit address). So the code should be fixed to look at the asce to see that type of table it is pointing at and then correctly follow the tables until it finds the data tables.
I've found the following share presentation helpful to understand the working of the ASCE:
www.kcats.org/csci/464/doc/s2825.pdf

**Comment by mfsysprog** [comment]
See 
https://github.com/torvalds/linux/blob/bf3e76289cd28b87f679cd53e26d67fd708d718a/arch/s390/include/asm/pgtable.h
for the ```#define _ASCE_TYPE_REGION3	0x04	/* region third table type	    */``` and ```#define _ASCE_TABLE_LENGTH	0x03	/* region table length		    */```

**Comment by mfsysprog** [comment]
After some more research I not sure anymore if it is a kernel issue. My idea is the following: The linux part allocates virtual memory to store the object, given the size (in pages) it received from the SCCB_HWL_TYPE_INFO call. It then has to build the neccesary ASCE structure and associated address translation tables for the service request to be able to locate those pages. If it is somewhere in the 2G - 4G memory region in the linux virtual memory for example, it requires a region third table as a start to be able to reach those pages. However, the first entry is not neccesarily the one to be a valid entry (the tables will only address the pages that are allocated to handle this request) and currently the code seems to assume it will get a segment table and that the address space should be filled from the bottom up. So whenever it finds the first invalid segment entry it will goto eof. So my though is that if the code is changed to identify the type of entry table and that it will scan the whole table for valid entries and then walks the next table, that it will then eventually find the actual pages that were allocated.

**Comment by mfsysprog** [comment]
I have succesfully managed to load the file! It is indeed as I expected in my last comment, the code has to correctly walk the region/segment/page tables to correctly identify which pages should be loaded. I have to change the code a bit to make sure that the ESA/390 mode still works ( I disregarded that trying to make a fix ), do some more testing and I will create a pull request in some time.
I did notice that a stat on the data will still show it as an empty file, but hexdump -C will show that it is actually loaded correctly.

**Comment by mfsysprog** [comment]
Even though my fix technically works, I did notice the following:
The ```sclp_sd_sync``` function in sclp_sd.c has the following comment:
``` * @page: Address of work page - must be below 2GB```

So perhaps the SCLP interface should 'officially' only handle 31 bit virtual addresses? 
As far as I know it is not publicly documented, so I cannot check unfortunately.
If this is indeed the case, then it is a bug in the linux code after all, since the memory allocation at the ```page = __get_free_page(GFP_KERNEL | GFP_DMA);``` should actually give a <2GB address (which should not be addressed by a region third table...)
The GFP_DMA should be set at 2GB according to this https://lwn.net/Articles/753273/

**Comment by mfsysprog** [comment]
One other thing I don't understand about the sclp_sd.c code is the following statement 
```sd_file->data_attr.read = data_read;```
Where does the ```data_read``` local variable come from? It is not passed in the function header and it is never allocated within the function. With my limited c knowledge I would say this cannot compile??

**Comment by Fish-Git** [comment]
> One other thing I don't understand about the sclp_sd.c code is the following statement:
> 
> ```
> sd_file->data_attr.read = data_read;
> ```
> 
> Where does the `data_read` local variable come from? It is not passed in the function header and it is never allocated within the function.

`data_attr.read` is a function pointer variable, and the statement is setting it to point to the `data_read` function:

* https://code.woboq.org/linux/linux/drivers/s390/char/sclp_sd.c.html#data_read

> With my limited C knowledge I would say this cannot compile?

Given that it's Linux kernel source code, I'm doubting that very much. &nbsp; `:)`

**Comment by mfsysprog** [comment]
I already figured that it was my limited C knowledge that was the issue here :-)
Looking at the function header for the data_read function:
```struct file * file, struct kobject * kobj, struct bin_attribute * attr, char * buffer, loff_t off, size_t size```
how does the code know what variables it must assign to each value from just:
```sd_file->data_attr.read = data_read;```
Does it automatically match local variables with the same variable type?

**Comment by mfsysprog** [comment]
Ah, I understand. It is not actually executing the function, it is just assigning the function pointer to ```sd_file->data_attr.read```

**Comment by wrljet** [comment]
Anybody with that pointer can call the read() function pointer within the structure.

The read() function pointer is defined in this structure:
This is what defines what arguments must be passed.

```
struct bin_attribute {
    struct attribute	attr;
    size_t size;
    void *private;
    ssize_t (*read)(struct file *, struct kobject *, struct bin_attribute *, char *, loff_t, size_t);
    ssize_t (*write)(struct file *, struct kobject *, struct bin_attribute *, char *, loff_t, size_t);
    int (*mmap)(struct file *, struct kobject *, struct bin_attribute *attr, struct vm_area_struct *vma);
};
```

For example, in function:

`sclp_sd_file *sclp_sd_file_create(const char *name, u8 di)`

It calls `sysfs_create_bin_file()` with that pointer, and also returns `sd_file` to its caller. 

It is those later callers who will know what variables to pass to the read function.
Without more research, I can't say where exactly it's called from.

https://www.geeksforgeeks.org/function-pointer-in-c/

**Comment by Fish-Git** [comment]
> I have successfully managed to load the file!
> 
> It is indeed as I expected in my last comment: the code has to correctly walk the region/segment/page tables to correctly identify which pages should be loaded.

Excellent!

The fix (commit b4c7fee88a0f3f7def2f0d2ea7663c55b5a704f6) has now been committed, so I am closing this issue.

Thank you for your effort working on this issue, Erik! Much appreciated!

(FYI: am closing Pull Request #337 as well.)

**Comment by bluecmd** [comment]
Very cool! Thanks folks!
